### DISCUSSION TITLE
Force layer to re-render

### DISCUSSION THREAD

**Discussion by ricardoekm** [visgl/deck.gl#6254] [Q&A] [answered]
We're using deck.gl with React+Redux, so every UI update goes through the React+Redux lifecycle (reducer, data comparison, components re-render, etc).

This usually works fine, however for more demanding tasks (e.g. drag an object) we see a drop in the FPS rate.

I wonder if there's a method available in the layer that would re-render it, without having to go all the way through the front-end stack. That way we could call it directly in the onDrag callback.

Thanks.

**Comment by Pessimistress** [comment, marked as answer]
You can directly access the underlying Deck instance like so:

```js
deckGLRef.current.deck.setProps({
  // partial props
})
```

However, What you are asking is to break the React programming pattern - the rendering updates to no longer be driven by prop changes. Given the description of your tech stack, I strongly recommend NOT to do so, because that puts your app's visual output in discrepancy with its data state, and may lead to bugs down the road.

We have built React applications that have similar functionality. My guess is that React itself is unlikely the cause of your performance problem. Some general advice without seeing your code:

- Intermediate state changes do not have to go through Redux. The common pattern is to store the intermediate position changes as objects are dragged in the state of a component, and only dispatch a change to Redux when drag ends.
- If you have multiple layers and/or child components passed to `<DeckGL>`, avoid creating new instances of them when only one of the layers require rerender. `useMemo` is your friend.
- For the layer whose object is dragged, the cost to update `data` is proportional to the length of the data array. You can invoke a partial data update for just one object with the [_dataDiff](https://deck.gl/docs/api-reference/core/layer#_datadiff) prop.

**Reply by vissharm** [reply]
@Pessimistress Regarding third point using _dataDiff prop, it will help me only in case of updating existing items to the layer. In this case buffers are not rewritten and I loved this work you guys done from performance perspective. But, what about if I am deleting or creating multiple object one by one or simultaneously.  As I read about _dataDiff it will rewrite all buffers if there is insertion or deletion and there will be whole recomputation for that. Do we have anything to help with insertion/deletion of items to the existing layer use-case, so current buffers are perserverd and new one's got created or deleted. I don't have much knowledge around GPU internals so this question might seems foolish but this is where I am stuck. We have use case where we can render max 8k paths and icons together. So performance matters alot, in this case I have to go for selective updates instead updating data prop and rerender everything again. Selective update using pubsub seems to be best way here but _datadiff will only help in case of updation of existing items in layer not for insertion and deletion.

In brief the **problem statement** is if I have 8000 icons in iconlayer and 8000 fibers in fiber layer and then I add one more icon or path to the respective layers... or delete one icon or path from the respective layers then how we can update the layers without recomputing unchanged data or existing data.

Anything I am missing here.. please let me know...

**Reply by Pessimistress** [reply]
The buffer is laid out based on the order of your data objects. There is no way around rewriting a large chunk of the buffer if you insist on shifting the objects around. If you want performance, then you need to arrange your data array in a way that allows for partial rewrite to work.

Instead of inserting an item into the middle of the array, always append it to the end. Use `_dataDiff` so that only the attributes of the new rows and evaluated.

Instead of deleting an item from the array, mark it as 'deleted', and set its width/size to 0. Use `_dataDiff` so that only the attributes of this item is rewritten.

GPU is dumb, you'll have a much easier time changing how you do things on the CPU, than trying to bend GPU to your fancy data structure.

**Comment by ricardoekm** [comment]
Great, thanks for the tips @Pessimistress!
